### PR TITLE
CsvImport: Proper escape handling

### PR DESCRIPTION
**Pull request by lanice**
This adds the ability to use any character as escape character, and proper escape handling ( #315 ).
Right now the escape character can be specified in the CsvConfig, but with the future json-meta file this will be specifiable there as well.

### CODE REVIEW

File: src/lib/import_export/csv_converter.hpp [hunk 1]
```diff
@@ -44,18 +44,22 @@ class AbstractCsvConverter {
...
-                   bool do_copy = last_char != config.escape || c != config.quote;
-                   // Set last_char to zero if the current character should not be copied
-                   // This is necessary because config.escape and config.quote are the same characters
-                   // and therefore a sequence of three quotes would trigger the condition above twice.
-                   last_char = do_copy ? c : 0;
+                 [&escaped, &config](const char c) {
+                   bool do_copy = true;
+
+                   // If escape character is found the first time, don't copy,
+                   // and set 'escaped' to true for the next character
```
Comment (mrks) [inline src/lib/import_export/csv_converter.hpp]: I like this approach.

File: src/lib/import_export/csv_parser.cpp [hunk 2]
```diff
@@ -136,7 +136,16 @@ bool CsvParser::_find_fields_in_chunk(std::string_view csv_content, const Table&
     from = pos + 1;
     const char elem = csv_content.at(pos);
 
-    if (elem == _csv_config.quote) in_quotes = !in_quotes;
+    // Make sure to "toggle" in_quotes ONLY if the quotes are not part of the string (i.e. escaped)
+    if (elem == _csv_config.quote) {
+      bool escaped = false;
+      if (elem != _csv_config.escape) {
```
Comment (mrks) [inline src/lib/import_export/csv_parser.cpp]: How does this check make sense if two lines above, we already found `elem` to be `_csv_config.quote`? Is this only about escaping with double quotes? If so, it needs a comment.

File: src/lib/import_export/csv_parser.cpp [hunk 3]
```diff
@@ -136,7 +136,16 @@ bool CsvParser::_find_fields_in_chunk(std::string_view csv_content, const Table&
     from = pos + 1;
     const char elem = csv_content.at(pos);
 
-    if (elem == _csv_config.quote) in_quotes = !in_quotes;
+    // Make sure to "toggle" in_quotes ONLY if the quotes are not part of the string (i.e. escaped)
+    if (elem == _csv_config.quote) {
+      bool escaped = false;
```
Comment (mrks) [inline src/lib/import_export/csv_parser.cpp]: The naming could be improved. What exactly is escaped?

File: src/lib/import_export/csv_parser.cpp (hunk 2 above)
Comment (lanice) [reply src/lib/import_export/csv_parser.cpp]: Basically it covers the case where the quote char is equal to the escape char (or rather where that's not the case). I'll update the clause to
`if (_csv_config.quote != _csv_config.escape)`
that should make that more clear.

**Comment by lanice** [comment]
> Is unescape called for every column? Shouldn't we only call it if we're in a string column?

It is called for every column. For non-string columns it is used to get rid of (potential) leading and trailing quotes. If there are non, it does nothing, so for CSVs where the usere doesn't use quotes around non-string fields, it does not have any effect.

**Pull request merged**